Deserialization issue with HMACGetSecretInput and HMACGetSecretOutput

Hi,

I have two question around extension in the option response.

1.) When having extensions in the `PublicKeyCredentialRequestOptions` or `PublicKeyCredentialCreationOptions`
The type `AuthenticationExtensionsClientInputs` is used to house the extensions.

Since all of these types are having Json related annotations from Jackson my assumption these are intended to be used for serialization.

However the type `HMACGetSecretInput` used here: https://github.com/webauthn4j/webauthn4j/blob/master/webauthn4j-core/src/main/java/com/webauthn4j/data/extension/client/AuthenticationExtensionsClientInputs.java#L43

Does not have a public constructor or any other annotation which would be required to make serializable by Jackson.
I also don't seem to find any specific deserializes written for it.

So creating something like this:
```
AuthenticationExtensionsClientInputs<AuthenticationExtensionClientInput> extensions = new AuthenticationExtensionsClientInputs.BuilderForAuthentication()
                .setAppid("testappid")
                .setAppidExclude("testappidexclude")
                .setHMACGetSecret(new HMACGetSecretInput(new byte[] {0}, new byte[] {1}))
                .setUvm(Boolean.TRUE)
                .set("myextension", "test")
                .build();

        ServerPublicKeyCredentialGetOptionsRequest serverPublicKeyCredentialGetOptionsRequest = ServerPublicKeyCredentialGetOptionsRequest.builder()
                .username("")
                .userVerification(USER_VERIFICATION_REQUIREMENT)
                .extensions(extensions)
                .build();
        String json = objectMapper.writeValueAsString(serverPublicKeyCredentialGetOptionsRequest);
```
Produces the following which is likely already wrong:
```
{"username":"","userVerification":"required","extensions":{"appid":"testappid","appidExclude":"testappidexclude","uvm":true,"credProps":null,"credentialProtectionPolicy":null,"enforceCredentialProtectionPolicy":null,"hmacCreateSecret":null,"hmacGetSecret":{"salt1":"AA","salt2":"AQ"},"myextension":"test"}}
```

Than parsing back this to an object by Jackson triggers the following exception:
```
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `com.webauthn4j.data.extension.HMACGetSecretInput` (no Creators, like default constructor, exist): cannot deserialize from Object value (no delegate- or property-based Creator)
```

Am I overlooking something or this is indeed a bug?


2.) I discovered several inconsistencies among the property names in `AuthenticationExtensionsClientInputs` and the IANA extension database: https://www.iana.org/assignments/webauthn/webauthn.xhtml
- `credentialProtectionPolicy` vs `credProtect` (?)
- `enforceCredentialProtectionPolicy` is this related to the `credProtect` or something else?
- `hmacCreateSecret` vs `hmac-secret`
- `hmacGetSecret` vs `hmac-secret`

Are those incorrect or simply some custom extension by the project and they have nothing to do with the ones on the IANA list?
(I didn't dive into the specification for each of those separately)

Version **0.28.5.RELEASE**
Using Jackson **2.18.2**

Thanks a lot,
Adam

To serialize or deserialize webauthn4j data types into JSON, use `JsonConverter` or Jackson `ObjectMapper` with `WebAuthnJSONModule`.
`WebAuthnJSONModule` contains custom serializers and deserializers for webauthn4j data types.

> parsing back this to an object by Jackson triggers the following exception:
>
```
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `com.webauthn4j.data.extension.HMACGetSecretInput` (no Creators, like default constructor, exist): cannot deserialize from Object value (no delegate- or property-based Creator)
```

This seems it is indeed a bug. Thank you for reporting!


> 2.) I discovered several inconsistencies among the property names in AuthenticationExtensionsClientInputs and the IANA extension database: https://www.iana.org/assignments/webauthn/webauthn.xhtml

The WebAuthn Extension Identifier is NOT the key name for `AuthenticationExtensionsClientInputs`.

For example, [Credential Protection](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#sctn-credProtect-extension) extention has `credProtect` as its extension identifier,
but please check the Client extension input definition closely:
```
partial dictionary AuthenticationExtensionsClientInputs {
  USVString credentialProtectionPolicy;
  boolean enforceCredentialProtectionPolicy = false;
};
```
It has two keys: `credentialProtectionPolicy` and `enforceCredentialProtectionPolicy`.

--- a/webauthn4j-core/src/main/java/com/webauthn4j/data/extension/HMACGetSecretInput.java
+++ b/webauthn4j-core/src/main/java/com/webauthn4j/data/extension/HMACGetSecretInput.java
@@ -16,6 +16,8 @@
 
 package com.webauthn4j.data.extension;
 
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
 import com.webauthn4j.util.ArrayUtil;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
@@ -24,10 +26,15 @@ import java.util.Arrays;
 
 public class HMACGetSecretInput {
 
+    @JsonProperty
     private final byte[] salt1;
+    @JsonProperty
     private final byte[] salt2;
 
-    public HMACGetSecretInput(@NotNull byte[] salt1, @Nullable byte[] salt2) {
+    @JsonCreator
+    public HMACGetSecretInput(
+            @NotNull @JsonProperty("salt1") byte[] salt1,
+            @Nullable @JsonProperty("salt2") byte[] salt2) {
         this.salt1 = ArrayUtil.clone(salt1);
         this.salt2 = ArrayUtil.clone(salt2);
     }

--- a/webauthn4j-core/src/main/java/com/webauthn4j/data/extension/HMACGetSecretOutput.java
+++ b/webauthn4j-core/src/main/java/com/webauthn4j/data/extension/HMACGetSecretOutput.java
@@ -16,6 +16,8 @@
 
 package com.webauthn4j.data.extension;
 
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
 import com.webauthn4j.util.ArrayUtil;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
@@ -24,10 +26,15 @@ import java.util.Arrays;
 
 public class HMACGetSecretOutput {
 
+    @JsonProperty
     private final byte[] output1;
+    @JsonProperty
     private final byte[] output2;
 
-    public HMACGetSecretOutput(@NotNull byte[] output1, @Nullable byte[] output2) {
+    @JsonCreator
+    public HMACGetSecretOutput(
+            @NotNull @JsonProperty("output1") byte[] output1,
+            @Nullable @JsonProperty("output2") byte[] output2) {
         this.output1 = ArrayUtil.clone(output1);
         this.output2 = ArrayUtil.clone(output2);
     }
@@ -47,7 +54,6 @@ public class HMACGetSecretOutput {
 
     @Override
     public boolean equals(Object o) {
-        if (this == o) return true;
         if (o == null || getClass() != o.getClass()) return false;
         HMACGetSecretOutput that = (HMACGetSecretOutput) o;
         return Arrays.equals(output1, that.output1) && Arrays.equals(output2, that.output2);

--- a/webauthn4j-core/src/test/java/com/webauthn4j/converter/AuthenticationExtensionsClientInputsConverterTest.java
+++ b/webauthn4j-core/src/test/java/com/webauthn4j/converter/AuthenticationExtensionsClientInputsConverterTest.java
@@ -18,10 +18,8 @@ package com.webauthn4j.converter;
 
 import com.webauthn4j.converter.exception.DataConversionException;
 import com.webauthn4j.converter.util.ObjectConverter;
-import com.webauthn4j.data.extension.client.AuthenticationExtensionClientInput;
-import com.webauthn4j.data.extension.client.AuthenticationExtensionsClientInputs;
-import com.webauthn4j.data.extension.client.FIDOAppIDExtensionClientInput;
-import com.webauthn4j.data.extension.client.RegistrationExtensionClientInput;
+import com.webauthn4j.data.extension.HMACGetSecretInput;
+import com.webauthn4j.data.extension.client.*;
 import org.junit.jupiter.api.Test;
 
 import static org.assertj.core.api.Assertions.assertThat;
@@ -32,13 +30,13 @@ class AuthenticationExtensionsClientInputsConverterTest {
 
     private final ObjectConverter objectConverter = new ObjectConverter();
 
-    private final AuthenticationExtensionsClientInputsConverter authenticationExtensionsClientInputsConverter = new AuthenticationExtensionsClientInputsConverter(objectConverter);
+    private final AuthenticationExtensionsClientInputsConverter target = new AuthenticationExtensionsClientInputsConverter(objectConverter);
 
 
     @SuppressWarnings("ConstantConditions")
     @Test
     void convertRegistrationExtensions_null_test() {
-        assertThatThrownBy(() -> authenticationExtensionsClientInputsConverter.convert(null)).isInstanceOf(DataConversionException.class);
+        assertThatThrownBy(() -> target.convert(null)).isInstanceOf(DataConversionException.class);
     }
 
     @Test
@@ -46,26 +44,30 @@ class AuthenticationExtensionsClientInputsConverterTest {
         AuthenticationExtensionsClientInputs.BuilderForAuthentication builder = new AuthenticationExtensionsClientInputs.BuilderForAuthentication();
         builder.setAppid("test");
         AuthenticationExtensionsClientInputs<AuthenticationExtensionClientInput> extensions = builder.build();
-        assertThat(authenticationExtensionsClientInputsConverter.convertToString(extensions)).isEqualTo("{\"appid\":\"test\"}");
+        assertThat(target.convertToString(extensions)).isEqualTo("{\"appid\":\"test\"}");
     }
 
     @SuppressWarnings("ConstantConditions")
     @Test
     void convertAuthenticationExtensionsToString_null_test() {
-        assertThatThrownBy(() -> authenticationExtensionsClientInputsConverter.convertToString(null)).isInstanceOf(DataConversionException.class);
+        assertThatThrownBy(() -> target.convertToString(null)).isInstanceOf(DataConversionException.class);
     }
 
     @Test
     void convert_test() {
-        String source = "{\"appid\":\"dummy\"}";
-        AuthenticationExtensionsClientInputs<AuthenticationExtensionClientInput> clientInputs = authenticationExtensionsClientInputsConverter.convert(source);
-        assertThat(clientInputs.getExtension(FIDOAppIDExtensionClientInput.class)).isEqualTo(new FIDOAppIDExtensionClientInput("dummy"));
+        String source = "{\"appid\":\"testappid\",\"appidExclude\":\"testappidexclude\",\"uvm\":true,\"hmacGetSecret\":{\"salt1\":\"AA\",\"salt2\":\"AQ\"},\"myextension\":\"test\"}";
+        AuthenticationExtensionsClientInputs<AuthenticationExtensionClientInput> clientInputs = target.convert(source);
+        assertThat(clientInputs.getExtension(FIDOAppIDExtensionClientInput.class)).isEqualTo(new FIDOAppIDExtensionClientInput("testappid"));
+        assertThat(clientInputs.getExtension(FIDOAppIDExclusionExtensionClientInput.class)).isEqualTo(new FIDOAppIDExclusionExtensionClientInput("testappidexclude"));
+        assertThat(clientInputs.getExtension(UserVerificationMethodExtensionClientInput.class)).isEqualTo(new UserVerificationMethodExtensionClientInput(true));
+        assertThat(clientInputs.getExtension(HMACSecretAuthenticationExtensionClientInput.class)).isEqualTo(new HMACSecretAuthenticationExtensionClientInput(new HMACGetSecretInput(new byte[] {0}, new byte[] {1})));
+        assertThat(clientInputs.getValue("myextension")).isEqualTo("test");
     }
 
     @Test
     void convert_null_test() {
         String source = "null";
-        AuthenticationExtensionsClientInputs<AuthenticationExtensionClientInput> clientInputs = authenticationExtensionsClientInputsConverter.convert(source);
+        AuthenticationExtensionsClientInputs<AuthenticationExtensionClientInput> clientInputs = target.convert(source);
         assertThat(clientInputs).isNull();
     }
 
@@ -73,7 +75,7 @@ class AuthenticationExtensionsClientInputsConverterTest {
     @Test
     void convert_with_invalid_extension_test() {
         String source = "{\"invalid\":\"\"}";
-        AuthenticationExtensionsClientInputs<RegistrationExtensionClientInput> extensions = authenticationExtensionsClientInputsConverter.convert(source);
+        AuthenticationExtensionsClientInputs<RegistrationExtensionClientInput> extensions = target.convert(source);
         assertThat(extensions.getUnknownKeys()).contains("invalid");
 
     }

--- a/webauthn4j-core/src/test/java/com/webauthn4j/converter/AuthenticationExtensionsClientOutputsConverterTest.java
+++ b/webauthn4j-core/src/test/java/com/webauthn4j/converter/AuthenticationExtensionsClientOutputsConverterTest.java
@@ -19,8 +19,12 @@ package com.webauthn4j.converter;
 
 import com.webauthn4j.converter.exception.DataConversionException;
 import com.webauthn4j.converter.util.ObjectConverter;
+import com.webauthn4j.data.extension.HMACGetSecretOutput;
+import com.webauthn4j.data.extension.UvmEntries;
+import com.webauthn4j.data.extension.client.*;
 import org.junit.jupiter.api.Test;
 
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
 class AuthenticationExtensionsClientOutputsConverterTest {
@@ -40,4 +44,14 @@ class AuthenticationExtensionsClientOutputsConverterTest {
         //noinspection ConstantConditions
         assertThatThrownBy(() -> target.convertToString(null)).isInstanceOf(DataConversionException.class);
     }
+
+    @Test
+    void convert_test() {
+        String source = "{\"appid\":true,\"uvm\":[],\"hmacGetSecret\":{\"output1\":\"AA\",\"output2\":\"AQ\"},\"myextension\":\"test\"}"; // "appidExclude":"testappidexclude"
+        AuthenticationExtensionsClientOutputs<AuthenticationExtensionClientOutput> clientOutputs = target.convert(source);
+        assertThat(clientOutputs.getExtension(UserVerificationMethodExtensionClientOutput.class)).isEqualTo(new UserVerificationMethodExtensionClientOutput(new UvmEntries()));
+        assertThat(clientOutputs.getExtension(HMACSecretAuthenticationExtensionClientOutput.class)).isEqualTo(new HMACSecretAuthenticationExtensionClientOutput(new HMACGetSecretOutput(new byte[] {0}, new byte[] {1})));
+        assertThat(clientOutputs.getValue("myextension")).isEqualTo("test");
+    }
+
 }

